File: conformance/golden-file-worldline/v1/identity_oracle.py
"""Independent Golden File Worldline identity construction and parsing."""

from __future__ import annotations

from dataclasses import dataclass
import subprocess

from corpus_protocol import (
    MAX_SOURCE_BYTES,
    ROOT,
    U16_MAX,
    U64_MAX,
    bounded_file_bytes,
    case_name,
    decimal,
    decoded_hex,
    fail,
    rows,
    safe_source_path,
    unique,
)

DATA_MAGIC = b"KEEP:BLOB:DATA\0\0"
ID_MAGIC = b"KEEP:BLOB:ID\0\0\0\0"
VERSION = 1
ALGORITHM = 1
BINARY_LENGTH = 59
MAX_TEXT_BYTES = 109
MAX_TOTAL_BYTES = 1_048_911
MAX_INVALID_TEXT_BYTES = 4_096

IDENTITY_COLUMNS = [
    "case",
    "source_kind",
    "source_parameter",
    "repetitions",
    "logical_length",
    "canonical_text",
    "canonical_binary_hex",
]
INVALID_TEXT_COLUMNS = ["case", "input_hex", "expected_outcome"]
REQUIRED_TEXT_OUTCOMES = {
    "keep.identity.input_too_long",
    "keep.identity.malformed_structure",
    "keep.identity.trailing_data",
    "keep.identity.invalid_scheme",
    "keep.identity.invalid_kind",
    "keep.identity.malformed_version",
    "keep.identity.unsupported_version",
    "keep.identity.unsupported_algorithm",
    "keep.identity.noncanonical_length",
    "keep.identity.length_overflow",
    "keep.identity.invalid_digest_length",
    "keep.identity.noncanonical_digest_case",
    "keep.identity.invalid_digest_alphabet",
}


@dataclass(frozen=True)
class IdentityFixture:
    content: bytes
    canonical_text: str
    canonical_binary: bytes


def source_bytes(
    row: dict[str, str],
    repetitions: int,
    declared_length: int,
) -> bytes:
    kind = row["source_kind"]
    parameter = row["source_parameter"]
    if kind == "empty-v1" and parameter == "-" and repetitions == 1:
        content = b""
    elif kind == "file-v1" and repetitions == 1:
        path = safe_source_path(parameter)
        if path.stat().st_size != declared_length:
            fail(
                f"{row['case']}: source size differs from its declaration"
            )
        content = bounded_file_bytes(
            path,
            MAX_SOURCE_BYTES,
            row["case"],
        )
    elif kind == "byte-ramp-v1" and parameter == "-" and repetitions > 0:
        generated_length = repetitions * 256
        if (
            generated_length > MAX_SOURCE_BYTES
            or generated_length != declared_length
        ):
            fail(
                f"{row['case']}: byte-ramp length is invalid or unbounded"
            )
        content = bytes(range(256)) * repetitions
    else:
        fail(f"{row['case']}: invalid source declaration")
    if len(content) != declared_length:
        fail(f"{row['case']}: source length mismatch")
    return content


def digest(payload: bytes) -> bytes:
    preimage = (
        DATA_MAGIC
        + VERSION.to_bytes(2, "big")
        + bytes([ALGORITHM])
    )
    preimage += payload + len(payload).to_bytes(8, "big")
    completed = subprocess.run(
        ["b3sum", "--no-names"],
        input=preimage,
        capture_output=True,
        check=False,
        timeout=30,
    )
    if completed.returncode != 0:
        fail(
            "b3sum failed: "
            f"{completed.stderr.decode(errors='replace').strip()}"
        )
    try:
        output = completed.stdout.decode("ascii")
    except UnicodeDecodeError as error:
        fail(f"b3sum returned non-ASCII output: {error}")
    if not output.endswith("\n") or output.count("\n") != 1:
        fail("b3sum returned noncanonical framing")
    return decoded_hex(output[:-1], "b3sum digest", 32)


def expected_text(length: int, identity_digest: bytes) -> str:
    return f"keep:blob:v1:blake3-256:{length}:{identity_digest.hex()}"


def expected_binary(length: int, identity_digest: bytes) -> bytes:
    return (
        ID_MAGIC
        + VERSION.to_bytes(2, "big")
        + bytes([ALGORITHM])
        + length.to_bytes(8, "big")
        + identity_digest
    )


def check_identities() -> dict[str, IdentityFixture]:
    identity_rows = rows(
        ROOT / "identities.tsv",
        "# keep.golden-file-worldline.identities/v1",
        IDENTITY_COLUMNS,
    )
    fixtures: dict[str, IdentityFixture] = {}
    total = 0
    for row in identity_rows:
        name = case_name(row["case"], "identities.tsv")
        unique(name, set(fixtures), "identities.tsv")
        repetitions = decimal(
            row["repetitions"],
            f"{name} repetitions",
            MAX_SOURCE_BYTES,
        )
        length = decimal(
            row["logical_length"],
            f"{name} logical_length",
            MAX_SOURCE_BYTES,
        )
        total += length
        if total > MAX_TOTAL_BYTES:
            fail("total materialized corpus exceeds bound")
        content = source_bytes(row, repetitions, length)
        identity_digest = digest(content)
        if row["canonical_text"] != expected_text(
            length,
            identity_digest,
        ):
            fail(f"{name}: canonical text mismatch")
        binary = decoded_hex(
            row["canonical_binary_hex"],
            f"{name} binary",
            BINARY_LENGTH,
        )
        if binary != expected_binary(length, identity_digest):
            fail(f"{name}: canonical binary mismatch")
        fixtures[name] = IdentityFixture(
            content,
            row["canonical_text"],
            binary,
        )
    required = {
        "empty",
        "small-text",
        "binary-ramp",
        "large-ramp",
        "state-a",
        "state-b",
    }
    if not required.issubset(fixtures):
        fail("identities.tsv: required v1 cases are absent")
    expected_b = (
        fixtures["state-a"].content[:6]
        + b"INSERTED\n"
        + fixtures["state-a"].content[6:]
    )
    if fixtures["state-b"].content != expected_b:
        fail("state-b is not the declared insertion into state-a")
    return fixtures


def canonical_decimal_bytes(value: bytes) -> bool:
    return value == b"0" or (
        bool(value)
        and ord("1") <= value[0] <= ord("9")
        and all(ord("0") <= digit <= ord("9") for digit in value)
    )


def text_outcome(encoded: bytes) -> str | None:
    try:
        encoded.decode("utf-8")
    except UnicodeDecodeError as error:
        fail(f"invalid-text.tsv: input is not UTF-8: {error}")
    if len(encoded) > MAX_TEXT_BYTES:
        return "keep.identity.input_too_long"
    fields = encoded.split(b":")
    if len(fields) < 6 or any(not field for field in fields[:6]):
        return "keep.identity.malformed_structure"
    if len(fields) > 6:
        return "keep.identity.trailing_data"
    scheme, kind, version, algorithm, length, identity_digest = fields
    if scheme != b"keep":
        return "keep.identity.invalid_scheme"
    if kind != b"blob":
        return "keep.identity.invalid_kind"
    if version != b"v1":
        number = version[1:] if version.startswith(b"v") else b""
        if not canonical_decimal_bytes(number) or int(number) > U16_MAX:
            return "keep.identity.malformed_version"
        return "keep.identity.unsupported_version"
    if algorithm != b"blake3-256":
        return "keep.identity.unsupported_algorithm"
    if not canonical_decimal_bytes(length):
        return "keep.identity.noncanonical_length"
    if int(length) > U64_MAX:
        return "keep.identity.length_overflow"
    if len(identity_digest) != 64:
        return "keep.identity.invalid_digest_length"
    for character in identity_digest:
        if ord("A") <= character <= ord("F"):
            return "keep.identity.noncanonical_digest_case"
        if character not in b"0123456789abcdef":
            return "keep.identity.invalid_digest_alphabet"
    return None


def check_invalid_text() -> None:
    invalid_rows = rows(
        ROOT / "invalid-text.tsv",
        "# keep.golden-file-worldline.invalid-text/v1",
        INVALID_TEXT_COLUMNS,
    )
    seen: set[str] = set()
    outcomes: set[str] = set()
    for row in invalid_rows:
        name = case_name(row["case"], "invalid-text.tsv")
        unique(name, seen, "invalid-text.tsv")
        encoded = decoded_hex(
            row["input_hex"],
            f"{name} input",
            MAX_INVALID_TEXT_BYTES,
            True,
        )
        observed = text_outcome(encoded)
        if observed is None or row["expected_outcome"] != observed:
            fail(
                f"{name}: expected text outcome does not match "
                f"{observed!r}"
            )
        outcomes.add(observed)
    if not REQUIRED_TEXT_OUTCOMES.issubset(outcomes):
        fail(
            "invalid-text.tsv: required v1 outcome coverage is absent"
        )

File: xtask/src/golden_file_worldline/identity_oracle.rs
//! This module owns independent identity materialization, hashing, and vectors.

use std::collections::{BTreeMap, BTreeSet};

use super::canonical_value::{EmptyHex, case_name, decimal, decoded_hex, unique};
use super::corpus_protocol::{MAX_SOURCE_BYTES, TableRow};
use super::{Corpus, GoldenError};

pub(super) const ALGORITHM: u8 = 1;
pub(super) const BINARY_LENGTH: usize = 59;
pub(super) const ID_MAGIC: [u8; 16] = *b"KEEP:BLOB:ID\0\0\0\0";
pub(super) const VERSION: u16 = 1;

const DATA_MAGIC: [u8; 16] = *b"KEEP:BLOB:DATA\0\0";
const MAX_TOTAL_BYTES: u64 = 1_048_911;
const IDENTITY_COLUMNS: [&str; 7] = [
    "case",
    "source_kind",
    "source_parameter",
    "repetitions",
    "logical_length",
    "canonical_text",
    "canonical_binary_hex",
];

pub(super) struct IdentityFixture {
    pub(super) content: Vec<u8>,
    pub(super) canonical_text: String,
    pub(super) canonical_binary: Vec<u8>,
}

pub(super) fn check_identities(
    corpus: &Corpus,
) -> Result<BTreeMap<String, IdentityFixture>, GoldenError> {
    let rows = corpus.rows(
        "identities.tsv",
        "# keep.golden-file-worldline.identities/v1",
        &IDENTITY_COLUMNS,
    )?;
    let mut fixtures = BTreeMap::new();
    let mut seen = BTreeSet::new();
    let mut total = 0_u64;
    for row in rows {
        let name = case_name(row.field("case")?, "identities.tsv")?.to_owned();
        unique(&name, &mut seen, "identities.tsv")?;
        let (fixture, length) = identity_fixture(corpus, &row, &name)?;
        total = total
            .checked_add(length)
            .ok_or_else(|| GoldenError::violation("total materialized corpus overflow"))?;
        if total > MAX_TOTAL_BYTES {
            return Err(GoldenError::violation(
                "total materialized corpus exceeds bound",
            ));
        }
        if fixtures.insert(name.clone(), fixture).is_some() {
            return Err(GoldenError::violation(format!(
                "identities.tsv: admitted identifier was displaced {name:?}"
            )));
        }
    }
    check_required_identities(&fixtures)?;
    check_state_insertion(&fixtures)?;
    Ok(fixtures)
}

fn identity_fixture(
    corpus: &Corpus,
    row: &TableRow,
    name: &str,
) -> Result<(IdentityFixture, u64), GoldenError> {
    let repetitions = source_number(row, "repetitions", name)?;
    let length = source_number(row, "logical_length", name)?;
    let content = source_bytes(corpus, row, repetitions, length)?;
    let identity_digest = digest(&content)?;
    let canonical_text = expected_text(length, &identity_digest);
    if row.field("canonical_text")? != canonical_text {
        return Err(GoldenError::violation(format!(
            "{name}: canonical text mismatch"
        )));
    }
    let binary = decoded_hex(
        row.field("canonical_binary_hex")?,
        &format!("{name} binary"),
        BINARY_LENGTH,
        EmptyHex::Refuse,
    )?;
    if binary != expected_binary(length, &identity_digest) {
        return Err(GoldenError::violation(format!(
            "{name}: canonical binary mismatch"
        )));
    }
    Ok((
        IdentityFixture {
            content,
            canonical_text,
            canonical_binary: binary,
        },
        length,
    ))
}

fn source_number(row: &TableRow, field: &'static str, name: &str) -> Result<u64, GoldenError> {
    let maximum = u64::try_from(MAX_SOURCE_BYTES).map_err(|source| {
        GoldenError::violation(format!(
            "{name}: source bound cannot be represented: {source}"
        ))
    })?;
    decimal(row.field(field)?, &format!("{name} {field}"), maximum)
}

pub(super) fn digest(payload: &[u8]) -> Result<blake3::Hash, GoldenError> {
    let length = u64::try_from(payload.len()).map_err(|source| {
        GoldenError::violation(format!("payload length cannot be represented: {source}"))
    })?;
    let mut hasher = blake3::Hasher::new();
    hasher.update(&DATA_MAGIC);
    hasher.update(&VERSION.to_be_bytes());
    hasher.update(&[ALGORITHM]);
    hasher.update(payload);
    hasher.update(&length.to_be_bytes());
    Ok(hasher.finalize())
}

pub(super) fn expected_text(length: u64, identity_digest: &blake3::Hash) -> String {
    format!(
        "keep:blob:v1:blake3-256:{length}:{}",
        identity_digest.to_hex()
    )
}

fn source_bytes(
    corpus: &Corpus,
    row: &TableRow,
    repetitions: u64,
    declared_length: u64,
) -> Result<Vec<u8>, GoldenError> {
    let name = row.field("case")?;
    let kind = row.field("source_kind")?;
    let parameter = row.field("source_parameter")?;
    let declared_usize = usize::try_from(declared_length).map_err(|source| {
        GoldenError::violation(format!(
            "{name}: source length cannot fit this platform: {source}"
        ))
    })?;
    let content = match (kind, parameter, repetitions) {
        ("empty-v1", "-", 1) => Vec::new(),
        ("file-v1", _, 1) => {
            let source = corpus.source_file(parameter)?;
            if source.len() != declared_length {
                return Err(GoldenError::violation(format!(
                    "{name}: source size differs from its declaration"
                )));
            }
            source.bounded_bytes(MAX_SOURCE_BYTES, name)?
        }
        ("byte-ramp-v1", "-", count) if count > 0 => byte_ramp(name, count, declared_usize)?,
        _ => {
            return Err(GoldenError::violation(format!(
                "{name}: invalid source declaration"
            )));
        }
    };
    if content.len() == declared_usize {
        Ok(content)
    } else {
        Err(GoldenError::violation(format!(
            "{name}: source length mismatch"
        )))
    }
}

fn byte_ramp(name: &str, repetitions: u64, declared_length: usize) -> Result<Vec<u8>, GoldenError> {
    let expected = repetitions
        .checked_mul(256)
        .ok_or_else(|| GoldenError::violation(format!("{name}: byte-ramp length overflow")))?;
    let maximum = u64::try_from(MAX_SOURCE_BYTES).map_err(|source| {
        GoldenError::violation(format!(
            "{name}: source bound cannot be represented: {source}"
        ))
    })?;
    let declared = u64::try_from(declared_length).map_err(|source| {
        GoldenError::violation(format!(
            "{name}: declared length cannot be represented: {source}"
        ))
    })?;
    if expected > maximum || expected != declared {
        return Err(GoldenError::violation(format!(
            "{name}: byte-ramp length is invalid or unbounded"
        )));
    }
    let mut content = Vec::with_capacity(declared_length);
    content.extend((u8::MIN..=u8::MAX).cycle().take(declared_length));
    Ok(content)
}

fn expected_binary(length: u64, identity_digest: &blake3::Hash) -> Vec<u8> {
    let mut encoded = Vec::with_capacity(BINARY_LENGTH);
    encoded.extend_from_slice(&ID_MAGIC);
    encoded.extend_from_slice(&VERSION.to_be_bytes());
    encoded.push(ALGORITHM);
    encoded.extend_from_slice(&length.to_be_bytes());
    encoded.extend_from_slice(identity_digest.as_bytes());
    encoded
}

fn check_required_identities(
    fixtures: &BTreeMap<String, IdentityFixture>,
) -> Result<(), GoldenError> {
    const REQUIRED: [&str; 6] = [
        "empty",
        "small-text",
        "binary-ramp",
        "large-ramp",
        "state-a",
        "state-b",
    ];
    if REQUIRED.iter().all(|name| fixtures.contains_key(*name)) {
        Ok(())
    } else {
        Err(GoldenError::violation(
            "identities.tsv: required v1 cases are absent",
        ))
    }
}

fn check_state_insertion(fixtures: &BTreeMap<String, IdentityFixture>) -> Result<(), GoldenError> {
    let state_a = fixtures
        .get("state-a")
        .ok_or_else(|| GoldenError::violation("identities.tsv: state-a is absent"))?;
    let state_b = fixtures
        .get("state-b")
        .ok_or_else(|| GoldenError::violation("identities.tsv: state-b is absent"))?;
    let (prefix, suffix) = state_a
        .content
        .split_at_checked(6)
        .ok_or_else(|| GoldenError::violation("state-a is shorter than insertion offset"))?;
    let expected_length = state_a
        .content
        .len()
        .checked_add(b"INSERTED\n".len())
        .ok_or_else(|| GoldenError::violation("state-b length overflow"))?;
    let mut expected = Vec::with_capacity(expected_length);
    expected.extend_from_slice(prefix);
    expected.extend_from_slice(b"INSERTED\n");
    expected.extend_from_slice(suffix);
    if state_b.content == expected {
        Ok(())
    } else {
        Err(GoldenError::violation(
            "state-b is not the declared insertion into state-a",
        ))
    }
}

#[cfg(test)]
mod tests;
